sample/buildSrc/src/main/kotlin/com/dropbox/sample/tasks/AffectedTasksPlugin.kt

```diff
@@ -0,0 +1,130 @@
+
+import com.dropbox.affectedmoduledetector.AffectedModuleDetector
+import com.dropbox.affectedmoduledetector.DependencyTracker
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.Task
+import org.gradle.api.tasks.testing.Test
+import java.util.*
+
+var TEST_TASK_TO_RUN_EXTENSION = "TestTasks"
+
+open class TestTasks {
+    var variantToTest = "debug"
+    val assembleAndroidTestTask = "assemble${variantToTest.capitalize()}AndroidTest"
+    val runAndroidTestTask = "connected${variantToTest.capitalize()}AndroidTest"
+    val jvmTest = "test${variantToTest.capitalize()}UnitTest"
+    val jvmTestBackup = "test"
+}
+
+/**
+ * unlike [AffectedTestsPlugin] which skips unaffected tests, this plugn instead creates a task and
+ * registers all affected test tasks. Advantage is speed in not needing to skip modules at a large scale
+ *
+ * Registers 3 tasks
+ * gradlew runAffectedUnitTests - runs jvm tests
+ * gradlew runAffectedAndroidTests - runs connected tests
+ * gradlew assembleAffectedAndroidTests - assembles but does not run on device tests, useful when working with device labs
+ */
+class AffectedTasksPlugin : Plugin<Project> {
+    var ANDROID_TEST_BUILD_VARIANT = "AndroidTest"
+
+    lateinit var testTasks: TestTasks
+    override fun apply(project: Project) {
+        project.extensions.add(TEST_TASK_TO_RUN_EXTENSION, TestTasks())
+        project.afterEvaluate {
+            val rootProject = project.rootProject
+            testTasks = requireNotNull(
+                project.extensions.findByName(TEST_TASK_TO_RUN_EXTENSION)
+            ) as TestTasks
+            registerAffectedTestTask(
+                "runAffectedUnitTests",
+                testTasks.jvmTest, testTasks.jvmTestBackup, rootProject,
+            )
+
+            registerAffectedAndroidTests(rootProject)
+            registerAffectedConnectedTestTask(rootProject)
+        }
+
+        filterAndroidTests(project)
+        filterUnitTests(project)
+    }
+
+    private fun registerAffectedTestTask(
+        taskName: String, testTask: String, testTaskBackup: String?,
+        rootProject: Project
+    ): Task {
+        val task = rootProject.tasks.register(taskName) { task ->
+            val paths = getAffectedPaths(testTask, testTaskBackup, rootProject)
+            paths.forEach { path ->
+                task.dependsOn(path)
+            }
+            task.enabled = paths.isNotEmpty()
+            task.onlyIf { paths.isNotEmpty() }
+        }
+        return task.get()
+    }
+
+    private fun getAffectedPaths(
+        task: String,
+        taskBackup: String?,
+        rootProject: Project
+    ):
+            Set<String> {
+        val paths = LinkedHashSet<String>()
+        rootProject.subprojects { subproject ->
+            val pathName = "${subproject.path}:$task"
+            val backupPath = "${subproject.path}:$taskBackup"
+            if (AffectedModuleDetector.isProjectProvided(subproject)) {
```

Comment:
If we are interested in performance, it seems be better to flip this around and have AffectedModuleDetector expose a list of affected projects then filter through that list, rather than start with every subproject and ask the affected module detector if a given subproject is affected

Note, maybe this isn't a performance concern

Reply:
This gets a bit tricky.  A method like this could only be called in an `afterEvaluate` block or in the execution phase.  However, we need this list to call `dependsOn` which has to run during the configuration phase.